### ISSUE TITLE
Toolchain: validate the project configuration

### ISSUE THREAD

**Issue by enten** [ctrlplusb/react-universally#287]
@ctrlplusb 

**I think that it will be good to validate the project configuration** (it's not a requirement).

Currently, the [project configuration](https://github.com/ctrlplusb/react-universally/blob/34e8ecc0ebeae9a0cab5642c60972d32f3e15f67/config/index.js) is imported with no validation step ([here](https://github.com/ctrlplusb/react-universally/blob/3067a4ae88d27d75bf4e75fcd01d5dc26f6dc475/tools/development/hotDevelopment.js#L11) and [there](https://github.com/ctrlplusb/react-universally/blob/3067a4ae88d27d75bf4e75fcd01d5dc26f6dc475/tools/scripts/build.js)).

If a developer remove flow and set a wrong value in the project configuration: the compilation may crash with an unclear error.

It will be a good thing if the project configuration was validated in `tools/index.js` before start the hot development and in `tools/scripts/build.js` before start a build.

I suggest to use [`prop-types`](https://www.npmjs.com/package/prop-types) to define schema in way similar to react's PropTypes validation.

> I made that choice in [`tension`](https://github.com/enten/tension/commit/7148b7fd899ccb9dd9a28ef41e53dc8ab2845acc) (check [here for schema declarations examples](https://github.com/enten/tension/blob/58f97788ed38ba3b4bee4459baa1bdee6d1477fd/lib/metafile.js#L46)).
> Please note that [I enhanced the validation with custom prop types](https://github.com/enten/tension/blob/58f97788ed38ba3b4bee4459baa1bdee6d1477fd/lib/utils/guard.js).

**Comment by ctrlplusb** [comment]
Great thanks for starting this discussion @enten 

It seems you have been busy during the holidays! :)

Sorry my responses have been fairly terse thusfar.  I hope to have more time soon.

I'd like to add a few other options for schema validation:

  - flow
  - tcomb-validation
  - joi

Flow may be the path of least resistance, but a part of me feels a bit uncomfortable in how I may be throwing flow down user's throats.  It's awesome, but it's still an opinionated addition to the project.  Also, what happens if a user runs `npm run flow:remove`?  I have been wondering if I should rip flow out into a feature branch...

Sorry I may be pulling away from the original point here. :)

**Comment by codepunkt** [comment]
@ctrlplusb Not sure if that helps getting in track to the original point here, but i work without flow and had it removed 😆

**Comment by enten** [comment]
> I'd like to add a few other options for schema validation:
> 
> flow
> tcomb-validation
> joi

I prefer [`lov`](https://www.npmjs.com/package/lov) than `joi` (because it's lighter). But both are limited for adding custom validation rules (and not support "function" type, because it's out of their scope).

**So, do you think that it will be useful to implement a validation step for the project config?**

----

> It seems you have been busy during the holidays! :)

I'm not on holidays sir (I'm working at the moment).

> Sorry my responses have been fairly terse thusfar. I hope to have more time soon.

Please, do not be.

> Sorry I may be pulling away from the original point here. :)

I understand that is an important decision and a serious topic.

**Comment by ctrlplusb** [comment]
I'm not 100% convinced as of yet.  I might defer to your experience with it over time within tension. :)

**Comment by enten** [comment]
> I'm not 100% convinced as of yet

You're not convinced to add a validation step? or using `prop-types` module?

Because `prop-types` is just a suggestion (tcomb-validation, joi or whatever validation module can do the job).

**I am interested by your opinion about adding a validation step** (not by the modules to do that).

**Comment by ctrlplusb** [comment]
I am not 100% convinced about adding a validation step.  I would like us to maybe wait a while on this one and then get some feedback from yourself in terms of how valuable it has been for your own projects.

My apprehension is that we would be adding an additional bit of code that would have to be kept in sync with the schema, which is likely to change for projects that fork the project.  I am not 100% sure that the extra overhead would be worth the validation step.  

But if you come back to me after some time of using it and say without a doubt it is useful then we could definitely experiment with it and bring it in as a feature. :)

**Comment by enten** [comment]
As you know, I have splitted the toolchain to allow compilers work in their own forked-process.

Make a validation of the project config in the main process prevents to launch compilers forked-process which will may crash.

The schemas which I have defined allow additional properties but are strict with the properties required by tension (even with properties which will be set dynamically by the compilers forked-process - through the use of the custom type [`T.reserved`](https://github.com/enten/tension/blob/58f97788ed38ba3b4bee4459baa1bdee6d1477fd/lib/utils/guard.js#L90)).

The schema can be stored beside the project configuration (in `config/schema.js`) for example. Or the configuration can specify a property to reference a function which will be used for validation (or not if the property missing).

It's just a suggestion my good sir: not a requirement 😄 

> But if you come back to me after some time of using it and say without a doubt it is useful

This kind of validation exists since the first versions. It's a requirement for me.

But you're 100% right! Better to let the users of react-universally tell us if it's useful. We will see that through some kind of issues.
